src/main/java/org/ice4j/ice/harvest/AbstractUdpListener.java

```diff
@@ -316,7 +336,12 @@ private void runInHarvesterThread()
         }
         while (true);
 
-        // TODO we are all done, clean up.
+        //now clean up and exit
+        for (MySocket candidateSocket : sockets.values())
+        {
+            candidateSocket.close();
```

Comment:
I've not tested this, other than the unit test that I supplied. You're right, this might be causing ConcurrentModificationExceptions. Do you see a more elegant solution than creating a new collection?